index.js

```diff
@@ -207,31 +208,36 @@ S3Storage.prototype._handleFile = function (req, file, cb) {
       params.ContentEncoding = opts.contentEncoding
     }
 
-    var upload = this.s3.upload(params)
+    var upload = new Upload({
+      client: this.s3,
+      params: params
+    })
 
     upload.on('httpUploadProgress', function (ev) {
       if (ev.total) currentSize = ev.total
     })
 
-    upload.send(function (err, result) {
-      if (err) return cb(err)
-
-      cb(null, {
-        size: currentSize,
-        bucket: opts.bucket,
-        key: opts.key,
-        acl: opts.acl,
-        contentType: opts.contentType,
-        contentDisposition: opts.contentDisposition,
-        contentEncoding: opts.contentEncoding,
-        storageClass: opts.storageClass,
-        serverSideEncryption: opts.serverSideEncryption,
-        metadata: opts.metadata,
-        location: result.Location,
-        etag: result.ETag,
-        versionId: result.VersionId
+    upload.done()
+      .then(function (result) {
+        cb(null, {
+          size: currentSize,
+          bucket: opts.bucket,
+          key: opts.key,
+          acl: opts.acl,
+          contentType: opts.contentType,
+          contentDisposition: opts.contentDisposition,
+          contentEncoding: opts.contentEncoding,
+          storageClass: opts.storageClass,
+          serverSideEncryption: opts.serverSideEncryption,
+          metadata: opts.metadata,
+          location: result.Location,
+          etag: result.ETag,
+          versionId: result.VersionId
+        })
+      })
+      .catch(function (err) {
+        cb(err)
       })
```

Comment:
This will end up calling the callback twice if the callback itself throws, which is not desirable. In addition, if the `cb(err)` throws that error will be swallowed and by the `catch` function and used to return a rejected promise, which will then be dangling.

Could we use [`util.callbackify`](https://nodejs.org/api/util.html#utilcallbackifyoriginal) here instead?

```suggestion
    util.callbackify(upload.done.bind(upload))(function (err, result) {
      if (err) return cb(err)

      cb(null, {
        size: currentSize,
        bucket: opts.bucket,
        key: opts.key,
        acl: opts.acl,
        contentType: opts.contentType,
        contentDisposition: opts.contentDisposition,
        contentEncoding: opts.contentEncoding,
        storageClass: opts.storageClass,
        serverSideEncryption: opts.serverSideEncryption,
        metadata: opts.metadata,
        location: result.Location,
        etag: result.ETag,
        versionId: result.VersionId
      })
    })
```